### PR TITLE
 Laboratory work #4, Natalia Kulakova - 22FPL1

### PR DESCRIPTION
**Pull request by kulak0va29**

### CODE REVIEW

File: lab_3_generate_by_ngrams/target_score.txt [hunk 1]
```diff
@@ -1 +1 @@
-10
+10
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/target_score.txt:1]: ?

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -98,6 +128,37 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+        if not isinstance(prompt, str) or not prompt:
+            raise ValueError("prompt must be a non-empty string")
+        if not isinstance(seq_len, int) or seq_len <= 0:
+            raise ValueError("seq_len must be a positive integer")
+
+        encoded_prompt = self._word_processor.encode(prompt)
+        if not encoded_prompt:
+            raise ValueError("Failed to encode the prompt")
+
+        while seq_len >= 1:
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:140]: Can you explain why you think while is better here than for?

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -98,6 +128,37 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            sorted_tokens = sorted(list(next_tokens.items()), key=lambda pair: (float(pair[1]), pair[0]), reverse=True)
+            probability = 0
+            for i, (_, prob) in enumerate(sorted_tokens):
+                probability += prob
+                if probability >= self._p_value:
+                    random_token = random.choice(sorted_tokens[:i + 1])
+                    encoded_prompt += (random_token[0],)
+                    break
+            else:
+                break
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:155]: it doesn't look good, does it?

**Review by marina-kaz — COMMENTED** [review]
Very good! Keep going
I like that you print messages along with error raises

File: lab_3_generate_by_ngrams/target_score.txt (hunk 1 above)
Comment (kulak0va29) [reply lab_3_generate_by_ngrams/target_score.txt:1]: I occasionally changed target score in the wrong lab

Comment (marina-kaz) [inline lab_3_generate_by_ngrams/target_score.txt:1]: then change it back please, you probably missed the empty line

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -56,6 +75,14 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            raise ValueError("Inappropriate input in _postprocess_decoded_text method")
+
+        sentences = ' '.join(decoded_corpus).split(self._end_of_word_token)
+        text = '. '.join([sentence.strip().capitalize() for sentence in sentences])
+        if text[-1] == ' ':
```
Comment (marina-kaz) [inline lab_4_fill_words_by_ngrams/main.py:83]: it is better to use strip for this

**Review by marina-kaz — COMMENTED** [review]
Looks nice!

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**